sklearn/_loss/loss.py

```diff
@@ -541,6 +541,10 @@ class AbsoluteError(BaseLoss):
     For a given sample x_i, the absolute error is defined as::
 
         loss(x_i) = |y_true_i - raw_prediction_i|
+
+    Note that the exact hessian = 0 almost everywhere (except at one point, therefore
+    differentiable = False). Optimization routines like in HGBT, however, need a
+    hessian > 0. Therefore, we assign 1.
```

Comment:
Is the "assign 1" correct in all cases?

From reading the code, if `sample_weight` is not `None` then `constant_hessian=False`. If `constant_hessian=False`, then the hessian are set to `empty`:

https://github.com/scikit-learn/scikit-learn/blob/188267212cb5459bfba947c9ece083c0b5f63518/sklearn/_loss/loss.py#L493-L500

Reply:
HGBT uses a *Newton boosting* scheme, i.e. a 2nd order Taylor approximation of the loss. By assigning hessians to 1 (or any positive constant), we are effectively doing a standard gradient boosting (gradient descent / first order, in function space) step.
Therefore, it is correct to the best of my knowledge. One of the rare references on this is https://arxiv.org/abs/1808.03064, Section 2.1.2 and 2.1.5.

Reply:
I'm trying to make sure the theory is consistent with the implementation. Currently, the code is not setting hessian to one when there are sample weights.

Reply:
Let’s see. Trees are fit to minimize $\sum_i h_i (tree(x_i) - g_i/h_i)^2$. Setting hessian $h$ to sample weights $w$ gives $\sum_i w_i (tree(x_i) - g_i/w_i)^2$. On top, the gradient is proportional to the weight, i.e. $g=w \tilde{g}$, such that we end up with $\sum_i w_i (tree(x_i) - \tilde{g}_i)^2$ 🎉

Reply:
Also, tests should cover/proof the correctness.